### PR TITLE
Ryanunderhill/customop dll

### PR DESCRIPTION
**Pull request by RyanUnderhill**
**Description**:  Previously OrtGetApi(version) was used to get the Ort API interface. The problem is that this still requires an imported function to call this one. Anyone also wanting to use this API will need to depend on OnnxRuntime.dll (or fail if it was statically linked).

To fix this, this change adds a new OrtApiBase struct. It has a 'GetApi(version)' method inside it. By moving it here, the OrtApiBase* can be passed to anyone wanting to use any version of the Ort APIs without taking a DLL dependency.

The second part of this change adds a 'RegisterCustomOpsLibrary' function to load a custom op DLL (which needs to be passed the OrtApiBase* to work).

### CODE REVIEW

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -202,7 +202,20 @@ typedef enum OrtMemType {
...
+struct OrtApi;
+typedef struct OrtApi OrtApi;
+
+struct OrtApiBase {
+  const OrtApi*(ORT_API_CALL* GetApi)(uint32_t version)NO_EXCEPTION;  // Pass in ORT_API_VERSION
+  const char*(ORT_API_CALL* GetVersionString)() NO_EXCEPTION;
+};
+typedef struct OrtApiBase OrtApiBase;
+
+ORT_EXPORT const OrtApiBase* ORT_API_CALL OrtGetApiBase() NO_EXCEPTION;  // Pass in ORT_API_VERSION
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]: Do we need this comment "// Pass in ORT_API_VERSION" on this line?

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 2]
```diff
@@ -312,6 +325,8 @@ struct OrtApi {
 	*/
   OrtStatus*(ORT_API_CALL* AddCustomOpDomain)(_Inout_ OrtSessionOptions* options, _In_ OrtCustomOpDomain* custom_op_domain)NO_EXCEPTION;
 
+  OrtStatus*(ORT_API_CALL* RegisterCustomOpsLibrary)(_Inout_ OrtSessionOptions* options, _In_ const char* library_path)NO_EXCEPTION;
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]: Should we use ORTCHAR_T* here for consistency?

Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_c_api.h]: The correct invocation of this function relies on a contract with the library being registered. Can we document this contract here or elsewhere?

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 3]
```diff
@@ -329,6 +325,24 @@ ORT_API_STATUS_IMPL(OrtApis::AddCustomOpDomain, _In_ OrtSessionOptions* options,
   API_IMPL_END
 }
 
+ORT_API_STATUS_IMPL(OrtApis::RegisterCustomOpsLibrary, _Inout_ OrtSessionOptions* options, _In_ const char* library_path) {
+  API_IMPL_BEGIN
+
+  void* handle;
+  Env::Default().LoadDynamicLibrary(library_path, &handle);
```
Comment (pranavsharma) [inline onnxruntime/core/session/onnxruntime_c_api.cc]: What's our policy around unloading? Do we also take responsibility to unload the library on the users behalf? Or give them a handle back to unload?

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (RyanUnderhill) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Nope, good catch

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 2 above)
Comment (RyanUnderhill) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: We should, but Env::LoadDynamicLibrary only takes const char * vs the wchar_t * on Windows.

Comment (RyanUnderhill) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Good idea, I put some comments in the header to describe it.

File: onnxruntime/core/session/onnxruntime_c_api.cc (hunk 3 above)
Comment (RyanUnderhill) [reply onnxruntime/core/session/onnxruntime_c_api.cc]: I added a final parameter 'library_handle' that returns the handle, and documented when to free it.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 2 above)
Comment (pranavsharma) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Yeah, we should have both versions of LoadDynamicLibrary just like we've for FileOpenRd and FileOpenWr. If we change it to ORTCHAR_T* after 1.0 release, will it be ABI incompatible?

**Pull request merged**